### PR TITLE
Make check_env assertions in regards to observation_space more actionable

### PR DESCRIPTION
**Pull request by FieteO**
## Description
Provide more actionable assertions when checking the observation_space, as requested in #1399 

In particular, add dedicated `AssertionError` for:
- a different shape of `obs` and `observation_space`
- a different dtype of `obs` and `observation_space`
- different lower bounds of `obs` and `observation_space`
- different upper bounds of `obs` and `observation_space`

Misc:
- address `make lint` warning about checking availability of module
- enhance `README` section about running tests

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

The current mode of failure is not telling much about why it fails.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: stable_baselines3/common/env_checker.py [hunk 1]
```diff
@@ -169,9 +169,16 @@ def _check_obs(obs: Union[tuple, dict, np.ndarray, int], observation_space: spac
...
+    assert observation_space.dtype == obs.dtype, f"The observation returned by the `{method_name}()` method does not match the data type of the given observation space. Expected: {observation_space.dtype}, actual: {obs.dtype}"
+    assert np.all(obs >= observation_space.low), f"The observation returned by the `{method_name}()` method does not match the lower bound of the given observation space. Expected: obs >= {np.min(observation_space.low)}, actual: {np.min(obs)}"
+    assert np.all(obs <= observation_space.high), f"The observation returned by the `{method_name}()` method does not match the upper bound of the given observation space. Expected: obs <= {np.min(observation_space.high)}, actual: {np.max(obs)}"
+
+
     assert observation_space.contains(
         obs
-    ), f"The observation returned by the `{method_name}()` method does not match the given observation space"
+    ), f"The observation returned by the `{method_name}()` method does not match the shape of the given observation space. Expected: `{1}`, actual: `{2}`"
```
Comment (FieteO) [inline stable_baselines3/common/env_checker.py:181]: This should most likely be reverted to the old message

**Comment by FieteO** [comment]
All working now, except for the following `pytype` exception:
```
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 175, in _check_obs: No attribute 'shape' on tuple [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 175, in _check_obs: No attribute 'shape' on dict [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 175, in _check_obs: No attribute 'shape' on int [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 180, in _check_obs: No attribute 'shape' on tuple [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 180, in _check_obs: No attribute 'shape' on dict [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 180, in _check_obs: No attribute 'shape' on int [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 183, in _check_obs: No attribute 'dtype' on tuple [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 183, in _check_obs: No attribute 'dtype' on dict [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 183, in _check_obs: No attribute 'dtype' on int [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 188, in _check_obs: No attribute 'dtype' on tuple [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 188, in _check_obs: No attribute 'dtype' on dict [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 188, in _check_obs: No attribute 'dtype' on int [attribute-error]
  In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]

For more details, see https://google.github.io/pytype/errors.html#attribute-error
ninja: build stopped: subcommand failed.
Leaving directory '.pytype'
make: *** [Makefile:8: pytype] Error 1
```
Any suggestions for fixing that? I don't have much experience with type checking in python...

File: stable_baselines3/common/env_checker.py [hunk 2]
```diff
@@ -169,6 +169,41 @@ def _check_obs(obs: Union[tuple, dict, np.ndarray, int], observation_space: spac
     elif _is_numpy_array_space(observation_space):
         assert isinstance(obs, np.ndarray), f"The observation returned by `{method_name}()` method must be a numpy array"
 
+    if isinstance(observation_space, spaces.Box):
+        # check obs dimensions, dtype and bounds
+        assert (
+            observation_space.shape == obs.shape
+        ), """\
```
Comment (araffin) [inline stable_baselines3/common/env_checker.py:176]: Please use f-strings, they also work with multi lines strings, or do something like:
```python
assert ..., (f"long message with {var1}" 
             f"rest of the message with {var2}")
```

(note: black will reformat it the correct way)

**Comment by FieteO** [comment]
> All working now, except for the following `pytype` exception when running `make type`:
> 
> ```
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 175, in _check_obs: No attribute 'shape' on tuple [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 175, in _check_obs: No attribute 'shape' on dict [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 175, in _check_obs: No attribute 'shape' on int [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 180, in _check_obs: No attribute 'shape' on tuple [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 180, in _check_obs: No attribute 'shape' on dict [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 180, in _check_obs: No attribute 'shape' on int [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 183, in _check_obs: No attribute 'dtype' on tuple [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 183, in _check_obs: No attribute 'dtype' on dict [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 183, in _check_obs: No attribute 'dtype' on int [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 188, in _check_obs: No attribute 'dtype' on tuple [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 188, in _check_obs: No attribute 'dtype' on dict [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> File "/home/fiete/Downloads/stable-baselines3/stable_baselines3/common/env_checker.py", line 188, in _check_obs: No attribute 'dtype' on int [attribute-error]
>   In Union[dict, int, numpy.ndarray[Any, numpy.dtype], tuple]
> 
> For more details, see https://google.github.io/pytype/errors.html#attribute-error
> ninja: build stopped: subcommand failed.
> Leaving directory '.pytype'
> make: *** [Makefile:8: pytype] Error 1
> ```
> 
> Any suggestions for fixing that? I don't have much experience with type checking in python...

Issue resolved

**Review by araffin — APPROVED** [review]
LGTM, thanks =)
(I refactored the tests and extended the checks to numpy arrays)

**Pull request merged**